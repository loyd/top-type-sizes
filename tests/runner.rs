use top_type_sizes::*;

macro_rules! test_sample {
    ($name:ident) => {
        #[test]
        fn $name() {
            static SAMPLE: &'static str =
                include_str!(concat!("samples/", stringify!($name), ".txt"));

            let refined_sample = reader::read(SAMPLE.as_bytes()).unwrap();
            let types = parser::parse(&refined_sample).unwrap();
            insta::assert_debug_snapshot!(format!("{}_types", stringify!($name)), types);

            // TODO: check the transformer.

            let output = formatter::format(types, &Default::default());
            println!("{output}");
            // TODO: make snapshots of output.
        }
    };
}

test_sample!(command);
test_sample!(control_flow);
test_sample!(alignment_enum);
test_sample!(tokio_udp);
test_sample!(several_types);
test_sample!(timex);
test_sample!(async_fn);
test_sample!(compiler_messages);

// TODO: add samples from rustc tests.
