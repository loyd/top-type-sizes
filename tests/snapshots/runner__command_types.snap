---
source: tests/runner.rs
expression: types
---
[
    Type {
        name: "std::sys::unix::process::process_common::Command",
        size: 208,
        align: 8,
        kind: Struct(
            StructType {
                items: [
                    Field(
                        Field {
                            kind: AdtField,
                            name: "env",
                            size: 32,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "cwd",
                            size: 16,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "groups",
                            size: 16,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "program",
                            size: 16,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "args",
                            size: 24,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "argv",
                            size: 24,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "closures",
                            size: 24,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "stdin",
                            size: 8,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "stdout",
                            size: 8,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "stderr",
                            size: 8,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "uid",
                            size: 8,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "gid",
                            size: 8,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "pgroup",
                            size: 8,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "program_kind",
                            size: 1,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "saw_nul",
                            size: 1,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "create_pidfd",
                            size: 1,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                ],
            },
        ),
        end_padding: Some(
            5,
        ),
    },
]
